src/Xamarin.Android.Build.Tasks/Tasks/GenerateJavaStubs.cs

```diff
@@ -110,73 +117,97 @@ void Run (DirectoryAssemblyResolver res)
+			bool hasExportReference = false;
+			bool haveMonoAndroid = false;
+			var allTypemapAssemblies = new HashSet<string> (StringComparer.OrdinalIgnoreCase);
+			var userAssemblies = new Dictionary<string, string> (StringComparer.OrdinalIgnoreCase);
 			foreach (var assembly in ResolvedAssemblies) {
-				if (bool.TryParse (assembly.GetMetadata (AndroidSkipJavaStubGeneration), out bool value) && value) {
+				bool value;
+				if (bool.TryParse (assembly.GetMetadata (AndroidSkipJavaStubGeneration), out value) && value) {
 					Log.LogDebugMessage ($"Skipping Java Stub Generation for {assembly.ItemSpec}");
 					continue;
 				}
+
+				bool addAssembly = false;
+				string fileName = Path.GetFileName (assembly.ItemSpec);
+				if (!hasExportReference && String.Compare ("Mono.Android.Export.dll", fileName, StringComparison.OrdinalIgnoreCase) == 0) {
+					hasExportReference = true;
+					addAssembly = true;
+				} else if (!haveMonoAndroid && String.Compare ("Mono.Android.dll", fileName, StringComparison.OrdinalIgnoreCase) == 0) {
+					haveMonoAndroid = true;
+					addAssembly = true;
+				} else if (MonoAndroidHelper.FrameworkAssembliesToTreatAsUserAssemblies.Contains (fileName)) {
+					if (!bool.TryParse (assembly.GetMetadata (AndroidSkipJavaStubGeneration), out value) || !value) {
+						string name = Path.GetFileNameWithoutExtension (fileName);
+						if (!userAssemblies.ContainsKey (name))
+							userAssemblies.Add (name, assembly.ItemSpec);
+						addAssembly = true;
+					}
+				}
+
+				if (addAssembly) {
+					if (!allTypemapAssemblies.Contains (assembly.ItemSpec))
+						allTypemapAssemblies.Add (assembly.ItemSpec);
+				}
+
 				res.Load (assembly.ItemSpec);
 			}
 
-			// However we only want to look for JLO types in user code
-			List<string> assemblies = new List<string> ();
+			// However we only want to look for JLO types in user code for Java stub code generation
 			foreach (var asm in ResolvedUserAssemblies) {
 				if (bool.TryParse (asm.GetMetadata (AndroidSkipJavaStubGeneration), out bool value) && value) {
 					Log.LogDebugMessage ($"Skipping Java Stub Generation for {asm.ItemSpec}");
 					continue;
 				}
-				if (!assemblies.All (x => Path.GetFileName (x) != Path.GetFileName (asm.ItemSpec)))
-					continue;
-				Log.LogDebugMessage ($"Adding {asm.ItemSpec} to assemblies.");
-				assemblies.Add (asm.ItemSpec);
-			}
-			foreach (var asm in MonoAndroidHelper.GetFrameworkAssembliesToTreatAsUserAssemblies (ResolvedAssemblies)) {
-				if (bool.TryParse (asm.GetMetadata (AndroidSkipJavaStubGeneration), out bool value) && value) {
-					Log.LogDebugMessage ($"Skipping Java Stub Generation for {asm.ItemSpec}");
-					continue;
-				}
-				if (!assemblies.All (x => Path.GetFileName (x) != Path.GetFileName (asm.ItemSpec)))
-					continue;
-				Log.LogDebugMessage ($"Adding {asm.ItemSpec} to assemblies.");
-				assemblies.Add (asm.ItemSpec);
+				if (!allTypemapAssemblies.Contains (asm.ItemSpec))
```

Comment:
Ditto here: we don't need `.Contains()`.

`.Contains()` needs to do "much" of the work of `.Add()` -- take the hash of the value, walk tables, etc. -- and `.Add()` is perfectly happy to do nothing if the value is already present, so just skip the `.Contains()`.